Assets/Tests/InputSystem/Plugins/UITests.cs

```diff
@@ -223,6 +245,194 @@ public IEnumerator MouseActions_CanDriveUI()
+        Assert.That(players[1].eventSystem.currentSelectedGameObject, Is.EqualTo(players[1].rightGameObject));
+
+        // Click right gameObject of player 0
+        InputSystem.QueueStateEvent(mouse, new MouseState { position = new Vector2(400, 100), buttons = 1 << (int)MouseButton.Left });
+        InputSystem.QueueStateEvent(mouse, new MouseState { position = new Vector2(400, 100), buttons = 0 });
+        InputSystem.Update();
+
+        foreach (var player in players)
+            player.eventSystem.InvokeUpdate();
+
+        Assert.That(players[0].eventSystem.currentSelectedGameObject, Is.EqualTo(players[0].rightGameObject));
+        Assert.That(players[1].eventSystem.currentSelectedGameObject, Is.EqualTo(players[1].rightGameObject));
+    }
+
+    [UnityTest]
+    [Category("Actions")]
+    // Check that two players can have separate UI and control it using separate gamepads, using MultiplayerEventSystem.
+    public IEnumerator JoystickActions_MultiplayerEventSystemKeepsPerPlayerSelection()
+    {
+        // Create devices.
+        var gamepads = new[] { InputSystem.AddDevice<Gamepad>(), InputSystem.AddDevice<Gamepad>() };
+        var players = new[] { CreateScene(0, 240), CreateScene(240, 480) };
+
+        for (var i = 0; i < 2; i++)
+        {
+            var map = new InputActionMap();
+            var moveAction = map.AddAction("move");
+            var submitAction = map.AddAction("submit");
+            var cancelAction = map.AddAction("cancel");
+
+            // Create bindings.
+            moveAction.AddBinding(gamepads[i].leftStick);
+            submitAction.AddBinding(gamepads[i].buttonSouth);
+            cancelAction.AddBinding(gamepads[i].buttonEast);
+
+            // Wire up actions.
+            players[i].uiModule.move = new InputActionProperty(moveAction);
+            players[i].uiModule.submit = new InputActionProperty(submitAction);
+            players[i].uiModule.cancel = new InputActionProperty(cancelAction);
+
+            players[i].leftChildReceiver.moveTo = players[i].rightGameObject;
+            players[i].rightChildReceiver.moveTo = players[i].leftGameObject;
+
+            // Enable the whole thing.
+            map.Enable();
+        }
+
+        // We need to wait a frame to let the underlying canvas update and properly order the graphics images for raycasting.
+        yield return null;
+
+        Assert.That(players[0].eventSystem.currentSelectedGameObject, Is.EqualTo(players[0].leftGameObject));
+        Assert.That(players[1].eventSystem.currentSelectedGameObject, Is.EqualTo(players[1].leftGameObject));
+
+        // Reset initial selection
+        players[0].leftChildReceiver.Reset();
+        players[1].leftChildReceiver.Reset();
+
+        // Check Player 0 Move Axes
+        InputSystem.QueueDeltaStateEvent(gamepads[0].leftStick, new Vector2(1.0f, 0.0f));
+        InputSystem.Update();
```

Comment:
These days, `InputTestFixture` has some nice wrappers around this kind of stuff. Can just do

```
Set(gamepads[0].leftStick, new Vector2(1, 0));
```

Reply:
This throws a NullReferenceException.